server/src/test/java/org/opensearch/cluster/routing/OperationRoutingTests.java

```diff
@@ -626,14 +634,154 @@ public void testAdaptiveReplicaSelection() throws Exception {
+        ResponseCollectorService collector = new ResponseCollectorService(clusterService);
+        Map<String, Long> outstandingRequests = new HashMap<>();
+
+        GroupShardsIterator<ShardIterator> groupIterator = opRouting.searchShards(state,
+            indexNames, null, null, collector, outstandingRequests);
+        assertThat("One group per index shard", groupIterator.size(), equalTo(numIndices * numShards));
+
+        // Test that the shards use a round-robin pattern when there are no stats
+        assertThat(groupIterator.get(0).size(), equalTo(numReplicas + 1));
+
+        ShardRouting firstChoice = groupIterator.get(0).nextOrNull();
+        assertNotNull(firstChoice);
+        searchedShards.add(firstChoice);
+        selectedNodes.add(firstChoice.currentNodeId());
+
+        groupIterator = opRouting.searchShards(state, indexNames, null, null, collector, outstandingRequests);
+
+        assertThat(groupIterator.size(), equalTo(numIndices * numShards));
+        ShardRouting secondChoice = groupIterator.get(0).nextOrNull();
+        assertNotNull(secondChoice);
+        searchedShards.add(secondChoice);
+        selectedNodes.add(secondChoice.currentNodeId());
+
+        // All the shards should be ranked equally since there are no stats yet
+        assertTrue(selectedNodes.contains("node_b2"));
+
+        // Since the primary shards are divided randomly between node_a0 and node_a1
+        assertTrue(selectedNodes.contains("node_a0") || selectedNodes.contains("node_a1"));
+
+        // Now let's start adding node metrics, since that will affect which node is chosen. Adding more load to node_b2
+        collector.addNodeStatistics("node_a0", 1, TimeValue.timeValueMillis(50).nanos(), TimeValue.timeValueMillis(50).nanos());
+        collector.addNodeStatistics("node_a1", 20, TimeValue.timeValueMillis(100).nanos(), TimeValue.timeValueMillis(150).nanos());
+        collector.addNodeStatistics("node_b2", 40, TimeValue.timeValueMillis(250).nanos(), TimeValue.timeValueMillis(250).nanos());
+        outstandingRequests.put("node_a0", 1L);
+        outstandingRequests.put("node_a1", 1L);
+        outstandingRequests.put("node_b2", 1L);
+
+        groupIterator = opRouting.searchShards(state, indexNames, null, null, collector, outstandingRequests);
+        // node_a0 or node_a1 should be the lowest ranked node to start
+        groupIterator.forEach(shardRoutings -> assertThat(shardRoutings.nextOrNull().currentNodeId(), containsString("node_a")));
+
+
+        // Adding more load to node_a0
+        collector.addNodeStatistics("node_a0", 10, TimeValue.timeValueMillis(200).nanos(), TimeValue.timeValueMillis(150).nanos());
+        groupIterator = opRouting.searchShards(state, indexNames, null, null, collector, outstandingRequests);
+        // node_a0 or node_a1 should be the lowest ranked node to start
+        groupIterator.forEach(shardRoutings -> assertThat(shardRoutings.nextOrNull().currentNodeId(), containsString("node_a")));
```

Comment:
is this blob needed - the comment on node_a0/node_a1 being lowest ranked still is confusing as this only changed the load selectively for a0? The real load is coming from in the next statement that follow